parflow_data_management/transport/tasks.py

```diff
@@ -0,0 +1,39 @@
+import json
+import io
+
+from asgiref.sync import async_to_sync
+from channels.layers import get_channel_layer
+from paramiko.rsakey import RSAKey
+from django.contrib.auth import get_user_model
+from xkcdpass import xkcd_password as xp
+from celery import shared_task
+
+from transport.models.key_pair import KeyPair
+from .consumers import compute_group_for_user
+
+@shared_task
+def generate_key_and_passphrase(user_id, key_pair_id):
+    # TODO: Load this once, instead of every time the function is called?
+    wordfile = xp.locate_wordfile()
+    wordlist = xp.generate_wordlist(wordfile=wordfile)
+    passphrase = xp.generate_xkcdpassword(wordlist)
+
+    new_key = RSAKey.generate(bits=4096)
+    public_key = new_key.get_base64()
+    out_desc = io.StringIO()
+    new_key.write_private_key(out_desc, password=passphrase)
+
+    # Update existing model instance
+    key_pair = KeyPair.objects.get(pk=key_pair_id)
+    key_pair.owner = get_user_model().objects.get(pk=user_id)
+    key_pair.public_key = public_key
+    key_pair.private_key_encrypted = out_desc.getvalue()
+    key_pair.save()
```

Comment:
How does this work in a celery context?

Reply:
In terms of synchronisity / accessing the database safely?

Reply:
Yes, if this is run on a celery worker on another server outside django, outside a request. Maybe we can talk more about this in person.